### PR TITLE
CWE-643 XPathInjection on java

### PR DESCRIPTION
**Pull request by SpaceWhite**
XPath injection query exists available in C# and javascript language but not in java.
This pull request consist of XPath Injection query to detect the cases which user input match against xpath evaluate or compile function.

XPathInjection detects the following function
jaxp
```
xpath.evaluate(expression, doc);
xpath.compile(expression_str)
```

dom4j
```
document.selectSingleNode(expression_str);
document.selectNodes(expression_str);
```

### CODE REVIEW

**Comment by yo-h** [comment]
Thank you for your contribution and apologies for the delay in reviewing it.

In order to streamline our review process, we have recently added `experimental` directories for each language with a [policy](https://github.com/Semmle/ql/blob/master/docs/experimental.md) for accepting contributions into it.

If you move this contribution into the `experimental` directory, we'll be able to merge this pull request sooner, and make follow-up improvements in subsequent pull requests. This will not adversely affect the amount of the bounty award, but it will allow us to proceed more quickly with it.

**Comment by SpaceWhite** [comment]
@yo-h Moved it to experimental

**Comment by yo-h** [comment]
Please add a commit to [autoformat](https://help.semmle.com/codeql/codeql-for-vscode/reference/editor.html#autoformatting) the query. This is required by one of our PR checks.

File: java/ql/src/experimental/CWE-643/XPathInjection.qhelp [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+</p>
+</overview>
+
+<recommendation>
+<p>
+If user input must be included in an XPath expression, pre-compile the query and use variable
+references to include the user input.
+</p>
+<p>
+XPath injection can also be prevented by using XQuery
```
Comment (felicitymay) [inline java/ql/src/experimental/CWE-643/XPathInjection.qhelp]:
```suggestion
XPath injection can also be prevented by using XQuery.
```

File: java/ql/src/experimental/CWE-643/XPathInjection.qhelp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+</p>
+
+</recommendation>
+
+<example>
+<p>
+In the first, second, and third example, the code accepts a user and password specified by the user, and uses this
```
Comment (felicitymay) [inline java/ql/src/experimental/CWE-643/XPathInjection.qhelp]:
```suggestion
In the first, second, and third example, the code accepts a name and password specified by the user, and uses this
```

File: java/ql/src/experimental/CWE-643/XPathInjection.qhelp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+unvalidated and unsanitized value in an XPath expression. This is vulnerable to the user providing
+special characters or string sequences that change the meaning of the XPath expression to search
+for different values.
+</p>
+
+<p>
+In the forth example, the code utilizes setXPathVariableResolver which prevents XPath Injection
+</p>
+<p>
+The fifth example is dom4j xpath injection example
```
Comment (felicitymay) [inline java/ql/src/experimental/CWE-643/XPathInjection.qhelp]:
```suggestion
The fifth example is a dom4j Xpath injection example.
```

File: java/ql/src/experimental/CWE-643/XPathInjection.qhelp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+<example>
+<p>
+In the first, second, and third example, the code accepts a user and password specified by the user, and uses this
+unvalidated and unsanitized value in an XPath expression. This is vulnerable to the user providing
+special characters or string sequences that change the meaning of the XPath expression to search
+for different values.
+</p>
+
+<p>
+In the forth example, the code utilizes setXPathVariableResolver which prevents XPath Injection
```
Comment (felicitymay) [inline java/ql/src/experimental/CWE-643/XPathInjection.qhelp]:
```suggestion
In the fourth example, the code utilizes setXPathVariableResolver which prevents XPath Injection.
```

**Comment by SpaceWhite** [comment]
@felicitymay @yo-h Fixed

**Comment by felicitymay** [comment]
Thanks for your updates on this PR too. Text all LGTM 😄

**Pull request merged**